libindy/src/services/pool/state_proof/node.rs

```diff
@@ -237,6 +238,131 @@ impl Node {
         }
     }
 
+    fn _get_all_values<'a>(&'a self, db: &'a TrieDB) -> IndyResult<Vec<(Vec<u8>, String)>> {
+        trace!("Getting all values, cur node: {:?}", self);
+        match *self {
+            Node::Full(ref node) => {
+                let mut res = vec![];
+                for nibble in 0..16 {
+                    let next = &node.nodes[nibble];
+                    if let Some(next) = next {
+                        trace!("Checking nibble {}", nibble);
+                        let mut sub_res: Vec<(Vec<u8>, String)> = next._get_all_values(db)?;
+                        trace!("Got some values: {:?}", sub_res);
+                        sub_res = sub_res.into_iter().map(|(mut key, val)| {
+                            let mut key_new: Vec<u8> = vec![nibble as u8];
+                            key_new.append(&mut key);
```

Comment:
If I get it correct the key in final result will be restored in multiply steps.
If it correct statement about current implementation I suggest to track current path from the root and construct key for returned value at the moment of it detection